bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/BusinessPartnerService.kt

```diff
@@ -230,16 +235,57 @@ class BusinessPartnerService(
         )
     }
 
-    private fun updateSharingState(entity: BusinessPartner, stateDto: TaskClientStateDto) {
+    private fun updateSharingState(externalId: String, stateDto: TaskClientStateDto) {
+
+        val errorMessage = if (stateDto.processingState.errors.isNotEmpty()) stateDto.processingState.errors.joinToString(" // ") { it.description } else null
+        val errorCode = if (stateDto.processingState.errors.isNotEmpty()) BusinessPartnerSharingError.SharingProcessError else null
+
         sharingStateService.upsertSharingState(
             SharingStateDto(
                 BusinessPartnerType.ADDRESS,
-                entity.externalId,
+                externalId,
                 sharingStateType = orchestratorMappings.toSharingStateType(stateDto.processingState.resultState),
+                sharingErrorCode = errorCode,
+                sharingErrorMessage = errorMessage,
                 taskId = stateDto.taskId
             )
         )
+    }
+
+    @Scheduled(cron = "\${cleaningService.pollingCron:-}", zone = "UTC")
+    fun finishCleaningTask() {
+
+        val taskIds = sharingStateRepository.findBySharingStateType(SharingStateType.Pending)
+        val nonNullTaskIds = taskIds.mapNotNull { it.taskId }
+
+        val taskStates = orchestrationApiClient.goldenRecordTasks.searchTaskStates(TaskStateRequest(nonNullTaskIds))
+
+        var validBusinessPartner: List<BusinessPartner> = emptyList()
 
+        //Map and verification
+        taskStates.tasks.forEach { task ->
+
+            val relatedSharingState = taskIds.find { it.taskId == task.taskId }
```

Comment:
If you associate the sharing states before by taskId you don't need to search here each iteration. You could just access the map